[BE] typing for decorators - library

aorenste opened the pull request:
Test Plan: unit tests

Differential Revision: D62302678




cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @ezyang @malfet @xuzhao9 @gramster @albanD @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv

Skylion007 commented on torch/_export/passes/replace_quantized_ops_with_standard_ops_pass.py:
```diff
@@ -76,20 +76,20 @@ def get_dequantized(
-            zero_point,
-            qmin,
-            qmax,
+            scale,  # type: ignore[arg-type]
```
FYI, seems like quantize_per_tensor and dequantize_per_tensor has incorrect typing.

aorenste replied:
Agreed.  We should fix that eventually - but this PR is about annotating library.impl which is a bigger issue since that will strip the types off anything that is annotated with it (like dequantize_per_tensor).

Skylion007 commented on torch/library.py:
```diff
@@ -536,43 +553,95 @@ def wrap(f):
-@functools.singledispatch
-def impl(qualname, types, func=None, *, lib=None):
-    """Register an implementation for a device type for this operator.
+# @functools.singledispatch confuses mypy
```
What's the exact bug, functools.singledispatch should be support by mypy now?

aorenste replied:
The basic use is supported. The problem is that it's impossible to properly type this function without using overload - and it seems that using singledispatch + overload confuses mypy - it complains on the later use of @impl.register that the overloaded function has no attribute "register".

But looking at it I did realize I can make it simpler by only skipping the impl.register during TYPE_CHECKING.

Skylion007 replied:
@aorenste I think it should work? You may just need to switch the order of singledispatch and overload I think?

ezyang commented:
Deferring to @zou3519 for library review

aorenste committed:
[BE] typing for decorators - library (#138969)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/138969

Test Plan: unit tests

Differential Revision: D62302678

aorenste commented:
@pytorchbot merge
